docs/src/references/mqtt-api.md

```diff
@@ -0,0 +1,836 @@
+
+|Segment|Description|
+|---|----|
+|&lt;identifier&gt;|The [identifier](#group-identifier) (e.g. device/service) associated with the command.|
+|&lt;cmd_type&gt;|Command type. Each command can define its own payload schema to allow commands to have parameters related to the command's function.|
+|&lt;cmd_id&gt;|Unique command id which is unique for the command instance. e.g. `123456`, `d511a86cab95be81` etc.|
+
+### Command examples
+
+The following table details some example command types which are supported by thin-edge.
+
+|Command Type|Example Topic|
+|------|---|
+|software_list|`te/<identifier>/cmd/software_list/<cmd_id>`|
+|software_update|`te/<identifier>/cmd/software_update/<cmd_id>`|
+|config_snapshot|`te/<identifier>/cmd/config_snapshot/<cmd_id>`|
+|config_update|`te/<identifier>/cmd/config_update/<cmd_id>`|
+|firmware_update|`te/<identifier>/cmd/firmware_update/<cmd_id>`|
+|restart|`te/<identifier>/cmd/restart/<cmd_id>`|
+
+The command would be interpreted differently based on the target entity.
+For example, the `restart` could mean either a device restart or a service restart based on the target entity.
+
+### Examples: With default device/service topic semantics
+
+#### Command to main device
+
+Command to fetch the software list from the main device:
+
+```sh te2mqtt
+tedge mqtt pub -r 'te/device/main///cmd/software_list/123' '{
+  "status": "init"
+}'
+```
+
+The `status` value will transition from `init` to the final `successful` or `failed` via many intermediate states
+such as `validating`, `downloading`, `executing` etc.
+
+#### Command to child device
+
+Command to update the firmware of a child device:
+
+```sh te2mqtt
+tedge mqtt pub -r 'te/device/child01///cmd/firmware_update/123' '{
+  "status": "init",
+  "attempt": 1,
+  "name": "OpenWRT",
+  "version": "22.03",
+  "url":"http://127.0.0.1:8000/tedge/file-transfer/tedge-child/firmware_update/93d50a297a8c235",
+  "sha256":"c036cbb7553a909f8b8877d4461924307f27ecb66cff928eeeafd569c3887e29"
+}'
+```
+
+#### Command to a service
+
+Command to update the configuration of a service:
+
+```sh te2mqtt
+tedge mqtt pub -r 'te/device/main/service/collectd/cmd/config_update/123' '{
+  "status": "init",
+  "type": "collectd"
+  "path": "/etc/collectd/collectd.conf",
+  "url": "http://127.0.0.1:8000/tedge/file-transfer/collectd/config_update/collectd",
+}'
+```
+
+### Commands metadata
+
+For commands as well, additional command specific metadata can be registered as retained messages on the `<cmd_type>` topic.
+
+For example, the supported configuration list of the main device can be declared as follows:
+
+```sh te2mqtt
+tedge mqtt pub -r te/device/main///cmd/config_snapshot '{
+  "description": "Upload a configuration from the device",
+  "supportedTypes": ["mosquitto", "tedge", "collectd"]
+}'
+```
+
+*TODO*: Unified config list for both `config_snapshot` and `config_update` commands or support different lists for both?
+
+## Health check
```

Comment:
> By the way, I understood that the plan was to introduce some specific topic to publish health status in a similar way as for telemetry data. Why did you change your mind ?

Since the request was modelled as a command, I felt that the response also being a command response would feel more natural. So used, the empty `<cmd-id>` trick with trailing `/` for the same. But, yeah, it is broken with retained messages and I'm happy to move to dedicated `health/` topics.

Now that we've changed the health status messages to its own `health/` topic, I'm wondering why we're still sticking with a `cmd` for the request. Why not just `te/.../health/check` for the request as well? That way, we still get consistency.

Keeping it as a `cmd` just creates more inconsistency, like it is not a retained message and doesn't expect responses on the same topic, like all other commands. So, why confuse the user when we already have a dedicated `health/` section in the topic.

Reply:
I would argue the opposite in terms of "consistency". If you want a component to run some action (e.g. check the health), then publishing a command (`/cmd/health/check`) fits more inline with the whole concept (and upcoming operations workflow concept).

Reply:
I don't have a strong opinion for one or another scheme. Only some questions.

- Why promoting `health` as a category of the same level as measurements or commands? Wouldn't be `status` a better global category for things as `health` or `usage-statistics`? I would avoid to come with a new top-category each time we see something a bit different.
- Would the users find helpful to have the request and the response under the same level (either `health` or `cmd/health`)
- Would the users find confusing to have commands with a different behavior? This might indeed introduces corner cases for the mqtt-state machine supervisor (unless we use reserved names as `req` and `res`).
- If two kinds of command is confusing we can have a `req` channel as in `req/health-check`.

I think these questions should not block this PR which is a huge step forward, the main point being the systematic way to identify devices and services.

Reply:
> Why promoting health as a category of the same level as measurements or commands? Wouldn't be status a better global category for things as health or usage-statistics? I would avoid to come with a new top-category each time we see something a bit different.

Nice suggestion. I fully agree, having `/status/health` would be more logical as well as adding space to add different types of status messages for different purposes, e.g. `/status/communication` etc.

> Would the users find helpful to have the request and the response under the same level (either health or cmd/health)

I don't think so in this case. As generally users will already be subscribing to the health status topic. It is only in cases where they want to prompt the service to update it status that they need the commands, however the request to update the service status only makes sense for services which don't have an mqtt client, as generally the MQTT LWT message is usually a reliable enough. Plus publishing a message to update all service health statuses won't have a suitable response channel due to the broadcast nature of the request.

> Would the users find confusing to have commands with a different behavior? This might indeed introduces corner cases for the mqtt-state machine supervisor (unless we use reserved names as req and res).

I thought this was already covered by the operation workflow PR? e.g. a simple operation and then a operation controlled via a workflow. However I think we can cover such topics in that PR anyway.

Reply:
>> Would the users find confusing to have commands with a different behavior? This might indeed introduces corner cases for the mqtt-state machine supervisor (unless we use reserved names as req and res).

>I thought this was already covered by the operation workflow PR? e.g. a simple operation and then a operation controlled via a workflow. However I think we can cover such topics in that PR anyway.

Indeed, in the operation workflow PR, I proposed to use `te/.../cmd/health-check/req`. Here, my bad, I improvised `te/.../cmd/health/check`. The latter looks nicer but introduces an issue. How can the *generic* mechanism guess that `check` is not a command unique id? We can have few reserved words (as `req` and `res`) but not domain specific ones.

=> This issue has to be solved in the workflow PR.  This PR can assume that a check request is sent on `te/.../cmd/health/check`.

Reply:
> Wouldn't be `status` a better global category for things as `health` or `usage-statistics`?

+1 to that.

> Would the users find helpful to have the request and the response under the same level (either `health` or `cmd/health`)

 I also agree with @reubenmiller here for this specific case where the request must be decoupled from the health status messages, as those can be sent at will, without an explicit request as well.

> Would the users find confusing to have commands with a different behavior? This might indeed introduces corner cases for the mqtt-state machine supervisor (unless we use reserved names as req and res).

This is the point where I'm a bit against the idea of modelling the health check as a command, as the general expectation from commands are that they have a terminal state at the end. The state transitions might be different from one op to another, but they all have a terminal state. This is where the health check request is weird, in the sense that it is more like a dangling request that no one takes to the terminal state. I'd have modelled it as a command if it behaved like a command at least for the most part. This is what I tried with the first attempt where the `successful` and `failed` status responses where used to represent `up` and `down`. But here, we're not using the full notion of command beyond that initial prompt.

But... that being said, when we had the dedicated `/health/status` topic, it just felt natural to add a `health/check` counterpart as well. But, with the proposed `/status/<xyz>` topics for any kind of status updates, I'm not sure if a request counterpart like `check/<xyz>` makes sense. I'm assuming that the `cmd` counterpart would look something like `....../cmd/status/<xyz>`, overloading the `<cmd-id>` with the status type `<xyz>`.

I also agree that we wouldn't want to create many different ways of sending requests/prompts to a target either. So, I'm willing to go the `cmd` route, accepting the inconsistencies that comes with it.

Reply:
Ok. I pushed the minor change to the health status topic, `/status/health`.

We'll address any other problems in follow up PRs, so I'll resolve this discussion.